scheduler/planner.go

```diff
@@ -0,0 +1,348 @@
+package scheduler
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/distro"
+	"github.com/evergreen-ci/evergreen/model/task"
+)
+
+// UnitCache stores an unordered collection of schedulable units. The
+// Unit type holds one or more tasks, but is handled by the scheduler
+// as a single object. While the constituent tasks in a unit have an
+// order, the unit themselves are an intermediate abstraction for the
+// planner which represent task groups, tasks with their dependencies,
+// or the tasks from a single version
+type UnitCache map[string]*Unit
+
+// AddWhen wraps AddNew, and is a noop if the conditional is false.
+func (cache UnitCache) AddWhen(cond bool, id string, t task.Task) {
+	if !cond {
+		return
+	}
+
+	if existing, ok := cache[id]; ok {
+		existing.Add(t)
+		return
+	}
+
+	cache.Create(id, t)
+}
+
+// AddNew adds an entire unit to a cache with the specified ID. If the
+// cached item exists, AddNew extends the existing unit with the tasks
+// from the passed unit.
+func (cache UnitCache) AddNew(id string, unit *Unit) {
+	if existing, ok := cache[id]; ok {
+		for _, t := range unit.tasks {
+			existing.Add(t)
+		}
+
+		return
+	}
+
+	cache[id] = unit
+}
+
+func (cache UnitCache) Exists(key string) bool { _, ok := cache[key]; return ok }
+
+// Create makes a new unit around the existing task, caching it with
+// the specified key, and returning the resulting unit. If there is an
+// existing cache item with the specified ID, then Create extends that
+// unit with this task. In both cases, the resulting unit is returned
+// to the caller.
+func (cache UnitCache) Create(id string, t task.Task) *Unit {
+	if unit, ok := cache[id]; ok {
+		unit.Add(t)
+		return unit
+	}
+
+	unit := NewUnit(t)
+	cache.AddNew(id, unit)
+	return unit
+}
+
+// Export returns an unordered sequence of unique Units.
+func (cache UnitCache) Export() TaskPlan {
+	seen := StringSet{}
+	tpl := TaskPlan{}
+	for id := range cache {
+		if seen.Visit(cache[id].ID()) {
+			continue
+		}
+
+		if cache[id].distro == nil {
+			continue
+		}
+
+		tpl = append(tpl, cache[id])
+	}
+
+	return tpl
+}
+
+// Unit is a holder of a group of related tasks which should be
+// scheculded together. Typically these represent task groups, tasks,
```

Comment:
Will tasks in a task group be ordered if they are grouped in a version?

Reply:
talked about this in person, will file a ticket to do more of this in the dispatcher.

Reply:
Also I think that because task group tasks end up in two Units, and then the lower priority version (e.g. the one that isn't in the version is prioritized higher,) is the one that is acted upon, this is actually fine.